### ISSUE TITLE
xref to web page kills PDF, hangs HTML5 for a while

### ISSUE THREAD

**Issue by robander** [dita-ot/dita-ot#2862]
## Expected Behavior

One of my customers had this link in a topic -- it's missing `@scope="external"` but is otherwise normal and worked in the past:
```
<p>For more information about each edition, see the 
<xref format="html"
href="http://www-01.ibm.com/common/ssi/ShowDoc.wss?docURL=/common/ssi/rep_ca/0/897/ENUS217-240/index.html"
>IBM 
  Software Announcement</xref>
```

Expected: the link works. Based on the specification, processing default for links beginning with http, https, ftp, and mailto should be treated as external, so my customers shouldn't be forced to add it (though I tend to encourage adding it), and certainly shouldn't get failing builds because of it:
https://docs.oasis-open.org/dita/dita/v1.3/errata01/os/complete/part1-base/langRef/attributes/thescopeattribute.html

## Actual Behavior

With a simple one-topic map, and that link in the topic, HTML5 hangs for an extended period at `topicpull`.

In 2.5.4, the HTML5 build otherwise works locally on my Windows machine, but when I run it on a Linux machine I get this error from the `topic-fragment` step -- I don't know why only there: 
`Error: java.lang.IllegalArgumentException: Only file URI scheme supported: http://www-01.ibm.com/common/ssi/ShowDoc.wss?docURL=/common/ssi/rep_ca/0/897/ENUS217-240/index.html`

I don't have 3.0 on that machine to test HTML5 there. But locally, with PDF and `preprocess2`, that link now kills the build in the merge step:
```
map2pdf2:
[topic-merge] Processing C:\Users\IBM_AD~1\AppData\Local\Temp\temp20171204144909813\e17ddccb7429a132b18bfa84b28d3bed6f844f1a.ditamap
[topic-merge] Processing C:\Users\IBM_AD~1\AppData\Local\Temp\temp20171204144909813\b7e631acad95a349a712d265a6b2a2d6792ac95e.dita
[topic-merge] null
Error: Failed to run pipeline: Failed to process merged topics: Supplied element name is a zero-length string
```

## Possible Solution

Per the specification, treat the 4 known prefixes as "external" when no other scope is specified or inherited (should be relatively straightforward for `xref` as in this case but not sure how this will affect cascading in maps).

Apart from the crash, which is obviously a bug itself -- not treating this as external is also a bug because the spec says it should be treated as external; the spec also allows other prefixes to be treated as external, but implementing this as some sort of configuration option seems more like a new feature.

## Steps to Reproduce

Build this map to PDF or HTML5 for the different behaviors described:
[scopes.zip](https://github.com/dita-ot/dita-ot/files/1528742/scopes.zip)

**Comment by raducoravu** [comment]
I'm not sure if the specs did well in treating href's which begin with http or https as external when no scope is present. I've seen a number of CMSs work over HTTP/Webdav and preferring to have full paths to resources in the @href value. In my opinion having the missing scope equivalent to local scope would have been more clear, anyway too late now.

**Comment by robander** [comment]
Hmm, thanks for that comment ... I was already nervous about making that change because of the difficulty of cascading scope values, but that's another reason to be nervous. Most important issue here is to fix the crash though.

**Comment by raducoravu** [comment]
👍

**Comment by robander** [comment]
Doing some more debug on this one, in 2.5.4 -- `filter` task tries to read the referenced HTML file when I'm processing on Unix, but not when I'm on Windows (the single-topic map hangs for several seconds at `filter` on UNIX but flies by on Windows). It looks like on UNIX it ends up listed as a DITA file in the `.job.xml` file, which is why it hangs later on a step that tries to process DITA topics:

```
<file 
   src="http://www-01.ibm.com/common/ssi/ShowDoc.wss?docURL=/common/ssi/rep_ca/0/897/ENUS217-240/index.html" 
   uri="http://www-01.ibm.com/common/ssi/ShowDoc.wss?docURL=/common/ssi/rep_ca/0/897/ENUS217-240/index.html" 
   path="http:/www-01.ibm.com/common/ssi/ShowDoc.wss?docURL=/common/ssi/rep_ca/0/897/ENUS217-240/index.html" 
   result="http://www-01.ibm.com/common/ssi/ShowDoc.wss?docURL=/common/ssi/rep_ca/0/897/ENUS217-240/index.html"
   format="dita" 
   resource-only="true"></file>
```

**Comment by robander** [comment]
And some debugging with 3.0 PDF -- in `preprocess2`, the `topic-reader` step writes out my original DITA topic with a generated name in the temp directory (as expected), but also decides that the xref with `format="html"` is DITA and rewrites the URL. Wonder if this is related to the behavior I see in 2.5.4 where the link is re-interpreted as DITA even though it has `format="html"` ... here is the updated `xref` immediately after the `topic-reader` stage:
```
<xref 
  dita-ot:orig-format="html" 
  format="dita" 
  href="4ab5f12f73c8751e05a2b6806e53c21c24e5ded1.wss" 
  class="- topic/xref " 
  xtrf="file:/C:/DCS/2.5.3/test/P025158/whats_new_cxa.dita" 
  xtrc="xref:1;11:2">IBM Software Announcement</xref>
```

**Comment by raducoravu** [comment]
Related problem, if in the DITA Map you refer to a PDF file:

           <topicref navtitle="PDF TITLE" href="tasks/flowers.pdf" format="pdf"/>

when publishing to XHTML based outputs the ".job.xml" considers it to have format DITA:

        <file src="file:/D:/projects/eXml/test/dita/transform/tasks/flowers.pdf"
            uri="tasks/flowers.pdf" path="tasks\flowers.pdf"
            result="file:/D:/projects/eXml/test/dita/transform/tasks/flowers.pdf" format="dita"/>

and the filtering stage tries to process it:

             [filter] Processing file:/D:/projects/eXml/test/dita/transform/tasks/flowers.pdf to file:/D:/projects/eXml/test/dita/transform/temp/xhtml/tasks/flowers.pdf

**Comment by robander** [comment]
Further debugging - it looks like this is better in 3.0.1 and in the latest 3.0.2 dev code, but still very broken -- rather than the crash I get in 3.0, I'm just getting a broken link. The original XREF, as in the original description above, is:
`<xref format="html" href="http://www-01.ibm.com/common/ssi/ShowDoc.wss?docURL=/common/ssi/rep_ca/0/897/ENUS217-240/index.html">IBM Software Announcement</xref>`

In the temp dir, this is converted to:
`<xref format="html" href="4ab5f12f73c8751e05a2b6806e53c21c24e5ded1.wss" class="- topic/xref " xtrf="file:/C:/DCS/2.5.3/test/P025158/whats_new_cxa.dita" xtrc="xref:1;11:2"><?ditaot usertext?>IBM Software Announcement</xref>`

The value `href="4ab5f12f73c8751e05a2b6806e53c21c24e5ded1.wss"` is of course wrong, and is not in the `.job.xml` file anywhere (so, it could not be mapped back to the original file, even if the intent was to create a generated value here).

**Comment by robander** [comment]
Problem is here (at least for my original report, haven't double checked the PDF format item from Radu but guessing the same): https://github.com/dita-ot/dita-ot/blob/develop/src/main/java/org/dita/dost/writer/DitaWriterFilter.java#L149

It's checking if the current attribute is `href`, and if so -- and if scope is local or unspecified -- rewrite with `replaceHREF`. I'm going to try adding the format check in there as well.

**Comment by robander** [comment]
That works, but makes me wonder about the generally unpleasant question of relative paths to local non-DITA files. The `replaceHREF` function does two things; it both renames files when the temp file name scheme is in use, and it adjusts relative paths to ensure references are still valid.

So, is this going to cause trouble?

- If I have a topic inside `subdir/` and it has a reference such as `<xref format="html" href="local.html">`, and
- If I'm using the preprocess2 renaming scheme that renames topics and moves them into the root directory, and
- I'm using an output format like HTML5 where I want the files written back out to the original structure (because that local file reference might make sense in HTML5 but not in PDF)

I'm pretty sure I could _make_ it cause trouble by reusing that `xref` inside a copied / conref'ed section, but I'm not certain. If the files are written back out to the original directory / file name scheme, then the original reference should be just fine. Either way, when using `preprocess2`, it should be preferable to leave non-DITA file names alone (don't touch the path) rather than renaming to a temp value that won't ever exist.

**Comment by raducoravu** [comment]
One remark about the fix, shouldn't we check for something like "isDITA or isDITAMap" or in that part of the code it's impossible to refer to a DITA Map?

**Comment by robander** [comment]
Probably a good point - I'm not sure if this is running on map elements or just on topics, but it's at least theoretically possible to reference a map from `<xref>` so we should keep up with any changes there too.

**Comment by robander** [comment]
OK, latest commit checks for both "dita" and "ditamap" as format.

But, that won't work either -- because `preprocess2` also moves and gives hashed names to images. (In the latest DITA 1.3 errata in #2860 you can specify `@format` directly on images, but it's already technically possible to reference them on a `topicref` with `@format`). If I update my test case to put a topic in a subdirectory, and add an image _without_ `@format`, all is good; but if I add `format="png"` then this new exception means the `href` is left alone even while the image is given a hashed name within the map directory. We could add another exception ("or class matches `topic/image`") but that feels like a hack and wouldn't catch images referenced via `topicref`. 

I think it's time for me to cry out to @jelovirt for advice, to see if there is any more elegant solution here, but ... I'm starting to fear the answer.

**Comment by robander** [comment]
Currently the logic for that section (As shipped in 3.0) is roughly...

- If the attribute is `@conref` then update it to use the new hash file name for the topic/map
- Else if the attribute is `@href`
  - If there's no `@scope` or `scope="local"` then update it to use a new hash file name
- Else if the attribute is `@format` ... then if `isFormatDita(format)` but it's not equal to `dita` then set it to `dita` and set a `dita-ot` attribute to store the original format (which must have been something from the configuration)

I'm not sure how to adjust that middle condition so that it adjust what we need (DITA, DITAMAP, local images) but not what we don't? 

- Just checking `@format` for non-DITA won't work for images
- Checking for absolute path + non-DITA format seems a lot more reliable (it would catch all the conditions that are failing for me now) ... but there might still be local images with absolute paths that we want to generate? _(Thinking about an absolute CMS reference to an image? But I don't have experience with that.)_
- Absolute path + non-DITA format + not on `<image>`? That feels like a hack and again, it's theoretically possible to construct `<topicref href="/some.jpg" format="jpg">`?
- Maybe ... if absolute path + non-DITA format + `@scope` unspecified, then we leave it alone? This still feels really complicated, but ... it would fix all of the problems I have, while still allowing the edge case of topicref to work with `<topicref href="/some.jpg" format="jpg" scope="local">`

**Comment by robander** [comment]
Closed with #2877
